diff-2.7/debian/cmp.1
.\" cmp(1) by Peter Samuelson <psamuels@sampo.creighton.edu> 23 September 1998
.\"
.\" Originally derived from the Debian Linux 2.0 distribution, who got
.\" it from FreeBSD, but I rewrote and reformatted this so extensively
.\" (using the manual page from GNU grep as a sort of template since I
.\" don't know roff very well) that I don't feel the need to retain the
.\" UCB copyright notice.
.\"
.\" This documents cmp from GNU diffutils version 2.7, with the byte
.\" ranges extension by Peter Samuelson.  It is in the public domain.
..
.TH CMP 1 "1998 September 23" "GNU Project"
..
.SH NAME
cmp (GNU diffutils) \- compare two files or byte ranges
..
.SH SYNOPSIS
.B cmp
[-clsv] [-i NUM] [--help] [--print-chars] [--ignore-initial=NUM]
[--verbose] [--quiet] [--silent] [--version] -I FILE1 [FILE2 [RANGE1
[RANGE2]]]
..
.SH DESCRIPTION
.PP
The
.B cmp
utility compares two files of any type and writes the results to the
standard output.  By default,
.B cmp
is silent if the files are the same; if they differ, the byte and line
number at which the first difference occurred is reported.
.PP
In the output, bytes and lines are numbered beginning with one;
however, range inputs are zero-based; see below for details.  A
filename of
.I -
represents standard input.
.PP
The following options are available:
.PD 0
.TP
.B "-c, --print-chars"
Output the differing bytes as characters, rather than as octal
numbers.  Non-printable characters will be shown in \ooo form.
.TP
.B "-i NUM, --ignore-initial=NUM"
Ignore
.I NUM
initial characters from each file.  This is a synonym for specifying
.I "NUM NUM"
as the two
.I RANGE
arguments.
.TP
.B "-l, --verbose"
Print the byte number (decimal) and the differing byte values (octal)
for each difference.
.TP
.B "-s, --quiet, --silent"
Print nothing for differing files; return exit status only.
.TP
.B "-v, --version"
Print the
.B diffutils
version number.
.PD
..
.SH "BYTE RANGES"
.PP
The two optional arguments
.I RANGE1
and
.I RANGE2
represent byte ranges to compare within the files.  Each range can be
expressed in several ways:
.PD 0
.TP
.B M+N
Skip
.I M
bytes at the beginning of the input, then compare a maximum of
.I N
bytes.
.TP
.B "M-N, M,N"
Skip
.I M
bytes at the beginning of the input, and read between bytes
.I M
and
.IR N ,
which are both zero-based.
.PD
.PP
In either case, both
.I M
and
.I N
are optional and default to beginning and end of file, respectively.
In addition, they can be expressed in decimal, octal
.RI ( 0NNN )
or hexadecimal
.RI ( 0xNNN )
form.
..
.SH NOTES
The zero-based range numbers may seem inconsistent with
.B cmp
output, which is one-based; this is for compatibility with some
versions of
.B cmp
which allow "skip
.I N
bytes" parameters after the filenames; in this context, the
.I N
is zero-based.
..
.SH DIAGNOSTICS
.PP
The
.B cmp
utility exits with one of the following values:
.PD 0
.TP
.B 0
The files or byte ranges are identical.
.TP
.B 1
The files or byte ranges are different; this includes the case where
one file or range is identical to the first part of the other.  In the
latter case, if
.B \-s
has not been specified,
.B cmp
writes to standard output that EOF was reached in the shorter file.
.TP
.B >1
An error occurred.
.PD
.SH SEE ALSO
.BR diff (1),
.BR diff3 (1)
.SH STANDARDS
The
.B cmp
utility is expected to be
.IR "POSIX 1003.2" -compliant.


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



diff-2.7/cmp.c
/* cmp -- compare two files.
   Copyright (C) 1990, 1991, 1992, 1993, 1994 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.  */

/* Written by Torbjorn Granlund and David MacKenzie. */
/* "ranges" feeping creaturism hacked in by Peter Samuelson. */

#include "system.h"
#include <stdio.h>
#include "getopt.h"
#include "cmpbuf.h"

extern char const version_string[];

#if __STDC__ && defined (HAVE_VPRINTF)
void error (int, int, char const *, ...);
#else
void error ();
#endif
VOID *xmalloc PARAMS((size_t));

static int cmp PARAMS((void));
static off_t file_position PARAMS((int, off_t));
static size_t block_compare PARAMS((char const *, char const *));
static size_t block_compare_and_count PARAMS((char const *, char const *, long *));
static size_t block_read PARAMS((int, char *, size_t));
static void printc PARAMS((int, unsigned));
static void try_help PARAMS((char const *));
static void check_stdout PARAMS((void));
static void usage PARAMS((void));
static off_t scan_off_t PARAMS((char **));
static int parse_range PARAMS((off_t[], char const *));

/* Name under which this program was invoked.  */
char const *program_name;

/* Filenames of the compared files.  */
static char const *file[2];

/* Byte ranges we need to compare.  */
static off_t range[2][2] = { {0,-1}, {0,-1} };

/* File descriptors of the files.  */
static int file_desc[2];

/* Read buffers for the files.  */
static char *buffer[2];

/* Optimal block size for the files.  */
static size_t buf_size;

/* Output format:
   type_first_diff
     to print the byte number and line number of the first differing bytes
   type_all_diffs
     to print the (decimal) byte numbers and (octal) values of all differing bytes
   type_status
     to only return an exit status indicating whether the files differ */
static enum
  {
    type_first_diff, type_all_diffs, type_status
  } comparison_type;

/* Type used for fast comparison of several bytes at a time.  */
#ifndef word
#define word int
#endif

/* If nonzero, print values of bytes quoted like cat -t does. */
static int opt_print_chars;

static struct option const long_options[] =
{
  {"print-chars", 0, 0, 'c'},
  {"ignore-initial", 1, 0, 'i'},
  {"verbose", 0, 0, 'l'},
  {"silent", 0, 0, 's'},
  {"quiet", 0, 0, 's'},
  {"version", 0, 0, 'v'},
  {"help", 0, 0, 129},
  {0, 0, 0, 0}
};

static void
try_help (reason)
     char const *reason;
{
  if (reason)
    error (0, 0, "%s", reason);
  error (2, 0, "Try `%s --help' for more information.", program_name);
}

static void
check_stdout ()
{
  if (ferror (stdout))
    error (2, 0, "write error");
  else if (fclose (stdout) != 0)
    error (2, errno, "write error");
}

static void
usage ()
{
  printf ("Usage: %s [OPTION]... FILE1 [FILE2 [RANGE1 [RANGE2]]]\n",
	  program_name);
  printf ("%s", "\
  -c  --print-chars  Output differing bytes as characters.\n\
  -i N  --ignore-initial=N  Ignore differences in the first N bytes of input.\n\
  -l  --verbose  Output byte numbers and codes of all differing bytes.\n\
  -s  --quiet  --silent  Output nothing; yield exit status only.\n\
  -v  --version  Output version info.\n\
  --help  Output this help.\n\
RANGE is a range of bytes in either file, and can be specified as follows:
  M+N    skip M bytes then compare the next N bytes\n\
  M-N    compare bytes M through N (zero-based, since it's skipping M bytes)\n\
  M      skip M bytes and read until EOF\n\
  +N, -N read the first N bytes of the file\n\
All numbers can be expressed in decimal, hex or octal.\n\
If a FILE is `-' or missing, read standard input.\n");
}

int
main (argc, argv)
     int argc;
     char *argv[];
{
  int c, i, exit_status;
  struct stat stat_buf[2];

  initialize_main (&argc, &argv);
  program_name = argv[0];

  /* Parse command line options.  */

  while ((c = getopt_long (argc, argv, "ci:lsv", long_options, 0))
	 != EOF)
    switch (c)
      {
      case 'c':
	opt_print_chars = 1;
	break;

      case 'i':
	range[0][0] = range[1][0] = scan_off_t(&optarg);
	break;

      case 'l':
	comparison_type = type_all_diffs;
	break;

      case 's':
	comparison_type = type_status;
	break;

      case 'v':
	printf ("cmp - GNU diffutils version %s\n", version_string);
	exit (0);

      case 129:
	usage ();
	check_stdout ();
	exit (0);

      default:
	try_help (0);
      }

  if (optind == argc)
    try_help ("missing operand");

  file[0] = argv[optind++];
  file[1] = optind < argc ? argv[optind++] : "-";

  if (optind < argc && parse_range(range[0], argv[optind++]))
    try_help ("invalid byte range");

  if (optind < argc && parse_range(range[1], argv[optind++]))
    try_help ("invalid byte range");

  if (optind < argc)
    try_help ("extra operands");

  for (i = 0; i < 2; i++)
    {
      /* If file[1] is "-", treat it first; this avoids a misdiagnostic if
	 stdin is closed and opening file[0] yields file descriptor 0.  */
      int i1 = i ^ (strcmp (file[1], "-") == 0);

      /* Two files with the same name are identical.
	 But wait until we open the file once, for proper diagnostics.  */
      if (i && filename_cmp (file[0], file[1]) == 0)
	exit (0);

      file_desc[i1] = (strcmp (file[i1], "-") == 0
		       ? STDIN_FILENO
		       : open (file[i1], O_RDONLY, 0));
      if (file_desc[i1] < 0 || fstat (file_desc[i1], &stat_buf[i1]) != 0)
	{
	  if (file_desc[i1] < 0 && comparison_type == type_status)
	    exit (2);
	  else
	    error (2, errno, "%s", file[i1]);
	}
#if HAVE_SETMODE
      setmode (file_desc[i1], O_BINARY);
#endif
    }

  /* If the files are links to the same inode and have the same file position,
     they are identical.  */

  if (0 < same_file (&stat_buf[0], &stat_buf[1])
      && file_position (0, range[0][0]) == file_position (1, range[1][0]))
    exit (0);

  /* If output is redirected to the null device, we may assume `-s'.  */

  if (comparison_type != type_status)
    {
      struct stat outstat, nullstat;

      if (fstat (STDOUT_FILENO, &outstat) == 0
	  && stat (NULL_DEVICE, &nullstat) == 0
	  && 0 < same_file (&outstat, &nullstat))
	comparison_type = type_status;
    }

  /* If only a return code is needed,
     and if both input descriptors are associated with plain files,
     conclude that the files differ if they have different sizes.  */

  if (comparison_type == type_status
      && S_ISREG (stat_buf[0].st_mode)
      && S_ISREG (stat_buf[1].st_mode))
    {
      off_t s0 = stat_buf[0].st_size - file_position (0, range[0][0]);
      off_t s1 = stat_buf[1].st_size - file_position (1, range[1][0]);

      if (range[0][1] != -1)
	s0 = min(s0, range[0][1]);
      if (range[1][1] != -1)
	s1 = min(s1, range[1][1]);

      if ( ( range[0][1] != -1 && range[1][1] != -1 &&
	     range[0][1] != range[1][1] ) ||
	   max(0, s0) != max(0, s1) )
	exit (1);
    }

  /* Get the optimal block size of the files.  */

  buf_size = buffer_lcm (STAT_BLOCKSIZE (stat_buf[0]),
			 STAT_BLOCKSIZE (stat_buf[1]));

  /* Allocate buffers, with space for sentinels at the end.  */

  for (i = 0; i < 2; i++)
    buffer[i] = xmalloc (buf_size + sizeof (word));

  exit_status = cmp ();

  for (i = 0; i < 2; i++)
    if (close (file_desc[i]) != 0)
      error (2, errno, "%s", file[i]);
  if (exit_status != 0  &&  comparison_type != type_status)
    check_stdout ();
  exit (exit_status);
  return exit_status;
}

/* Compare the two files already open on `file_desc[0]' and `file_desc[1]',
   using `buffer[0]' and `buffer[1]'.
   Return 0 if identical, 1 if different, >1 if error. */

static int
cmp ()
{
  long line_number = 1;		/* Line number (1...) of first difference. */
  long char_number = range[0][0] + 1;
				/* Byte number (1...) in files of 1st difference. */
  size_t read0, read1;		/* Number of chars read from each file. */
  size_t first_diff;		/* Offset (0...) in buffers of 1st diff. */
  size_t smaller;		/* The lesser of `read0' and `read1'. */
  char *buf0 = buffer[0];
  char *buf1 = buffer[1];
  int ret = 0;
  int i;

  for (i = 0; i < 2; i++)
    if (range[i][0] && file_position (i, range[i][0]) == -1)
      {
	/* lseek failed; read and discard the ignored initial prefix.  */
	off_t ig = range[i][0];
	do
	  {
	    size_t r = read (file_desc[i], buf0, (size_t) min (ig, buf_size));
	    if (!r)
	      break;
	    if (r == -1)
	      error (2, errno, "%s", file[i]);
	    ig -= r;
	  }
	while (ig);
      }
  
  do
    {
      read0 = block_read (file_desc[0], buf0, buf_size);
      if (read0 == -1)
	error (2, errno, "%s", file[0]);
      read1 = block_read (file_desc[1], buf1, buf_size);
      if (read1 == -1)
	error (2, errno, "%s", file[1]);

      /* Insert sentinels for the block compare.  */

      buf0[read0] = ~buf1[read0];
      buf1[read1] = ~buf0[read1];

      /* If the line number should be written for differing files,
	 compare the blocks and count the number of newlines
	 simultaneously.  */
      first_diff = (comparison_type == type_first_diff
		    ? block_compare_and_count (buf0, buf1, &line_number)
		    : block_compare (buf0, buf1));

      char_number += first_diff;
      smaller = min (read0, read1);

      /* Reached the end of one of the ranges, so we succeed. */
      if ( (range[0][1] != -1 && char_number >= range[0][1]) ||
	   (range[1][1] != -1 && char_number >= range[1][1]) )
	return 0;

      if (first_diff < smaller)
	{
	  switch (comparison_type)
	    {
	    case type_first_diff:
	      /* See Posix.2 section 4.10.6.1 for this format.  */
	      printf ("%s %s differ: char %lu, line %lu",
		      file[0], file[1], char_number, line_number);
	      if (opt_print_chars)
		{
		  unsigned char c0 = buf0[first_diff];
		  unsigned char c1 = buf1[first_diff];
		  printf (" is %3o ", c0);
		  printc (0, c0);
		  printf (" %3o ", c1);
		  printc (0, c1);
		}
	      putchar ('\n');
	      /* Fall through. */
	    case type_status:
	      return 1;

	    case type_all_diffs:
	      do
		{
		  unsigned char c0 = buf0[first_diff];
		  unsigned char c1 = buf1[first_diff];
		  if (c0 != c1)
		    {
		      if (opt_print_chars)
			{
			  printf ("%6lu %3o ", char_number, c0);
			  printc (4, c0);
			  printf (" %3o ", c1);
			  printc (0, c1);
			  putchar ('\n');
			}
		      else
			/* See Posix.2 section 4.10.6.1 for this format.  */
			printf ("%6lu %3o %3o\n", char_number, c0, c1);
		    }
		  char_number++;
		  first_diff++;
		}
	      while (first_diff < smaller);
	      ret = 1;
	      break;
	    }
	}

      if (read0 != read1)
	{
	  if (comparison_type != type_status)
	    /* See Posix.2 section 4.10.6.2 for this format.  */
	    fprintf (stderr, "cmp: EOF on %s\n", file[read1 < read0]);

	  return 1;
	}
    }
  while (read0 == buf_size);
  return ret;
}

/* Compare two blocks of memory P0 and P1 until they differ,
   and count the number of '\n' occurrences in the common
   part of P0 and P1.
   Assumes that P0 and P1 are aligned at word addresses!
   If the blocks are not guaranteed to be different, put sentinels at the ends
   of the blocks before calling this function.

   Return the offset of the first byte that differs.
   Increment *COUNT by the count of '\n' occurrences.  */

static size_t
block_compare_and_count (p0, p1, count)
     char const *p0, *p1;
     long *count;
{
  word l;		/* One word from first buffer. */
  word const *l0, *l1;	/* Pointers into each buffer. */
  char const *c0, *c1;	/* Pointers for finding exact address. */
  long cnt = 0;		/* Number of '\n' occurrences. */
  word nnnn;		/* Newline, sizeof (word) times.  */
  int i;

  l0 = (word const *) p0;
  l1 = (word const *) p1;

  nnnn = 0;
  for (i = 0; i < sizeof (word); i++)
    nnnn = (nnnn << CHAR_BIT) | '\n';

  /* Find the rough position of the first difference by reading words,
     not bytes.  */

  while ((l = *l0++) == *l1++)
    {
      l ^= nnnn;
      for (i = 0; i < sizeof (word); i++)
	{
	  cnt += ! (unsigned char) l;
	  l >>= CHAR_BIT;
	}
    }

  /* Find the exact differing position (endianness independent).  */

  c0 = (char const *) (l0 - 1);
  c1 = (char const *) (l1 - 1);
  while (*c0 == *c1)
    {
      cnt += *c0 == '\n';
      c0++;
      c1++;
    }

  *count += cnt;
  return c0 - p0;
}

/* Compare two blocks of memory P0 and P1 until they differ.
   Assumes that P0 and P1 are aligned at word addresses!
   If the blocks are not guaranteed to be different, put sentinels at the ends
   of the blocks before calling this function.

   Return the offset of the first byte that differs.  */

static size_t
block_compare (p0, p1)
     char const *p0, *p1;
{
  word const *l0, *l1;
  char const *c0, *c1;

  l0 = (word const *) p0;
  l1 = (word const *) p1;

  /* Find the rough position of the first difference by reading words,
     not bytes.  */

  while (*l0++ == *l1++)
    ;

  /* Find the exact differing position (endianness independent).  */

  c0 = (char const *) (l0 - 1);
  c1 = (char const *) (l1 - 1);
  while (*c0 == *c1)
    {
      c0++;
      c1++;
    }

  return c0 - p0;
}

/* Read NCHARS bytes from descriptor FD into BUF.
   Return the number of characters successfully read.
   The number returned is always NCHARS unless end-of-file or error.  */

static size_t
block_read (fd, buf, nchars)
     int fd;
     char *buf;
     size_t nchars;
{
  char *bp = buf;

  do
    {
      size_t nread = read (fd, bp, nchars);
      if (nread == -1)
	return -1;
      if (nread == 0)
	break;
      bp += nread;
      nchars -= nread;
    }
  while (nchars != 0);

  return bp - buf;
}

/* Print character C, making unprintable characters
   visible by quoting like cat -t does.
   Pad with spaces on the right to WIDTH characters.  */

static void
printc (width, c)
     int width;
     unsigned c;
{
  register FILE *fs = stdout;

  if (! ISPRINT (c))
    {
      if (c >= 128)
	{
	  putc ('M', fs);
	  putc ('-', fs);
	  c -= 128;
	  width -= 2;
	}
      if (c < 32)
	{
	  putc ('^', fs);
	  c += 64;
	  --width;
	}
      else if (c == 127)
	{
	  putc ('^', fs);
	  c = '?';
	  --width;
	}
    }

  putc (c, fs);
  while (--width > 0)
    putc (' ', fs);
}

/* Position file I to `pos' bytes from its initial position,
   and yield its new position.  Don't try more than once.  */

static off_t
file_position (i, pos)
     int i;
     off_t pos;
{
  static int positioned[2];
  static off_t position[2];

  if (! positioned[i] || position[i] != pos)
    {
      positioned[i] = 1;
      position[i] = lseek (file_desc[i], pos, SEEK_CUR);
    }
  return position[i];
}

/* scan_off_t(): sort of a specialized sscanf() that reads "%i" (dec,
 * hex, or oct) into an off_t.  This is in case (a) `off_t' is a bigger
 * type than `long' and (b) our compiler doesn't like `long long'.
 * Returns the integer (sorry, the "off_t") and leaves the string
 * pointer just past the end of the number it read.
 */
static off_t
scan_off_t (str)
     char **str;
{
  off_t n;
  unsigned int i, base = 10;
  char *s = *str;

  if(! *s)
    return 0;

  if(*s == '0')
    {
      if(*(++s) == 'x' || *s == 'X')
	{
	  base = 16;
	  s++;
	}
      else
	base=8;
    }

  for(n=0; *s; s++)
    {
      i = *s - '0';
      i -= ('A'-'0'-10) * (i > 9);
      i -= ('a'-'A') * (i > 'a'-'A');
      if(i > base)
	break;
      n = n*base+i;
    }
  *str = s;
  return n;
}

/* parse_range: given a char* and an off_t[2], return a byte range in
 * the off_t[2] in the form {start,nbytes}.  For example, given
 * "19-33", return {19,15}.  Given "22+3" return {22,3}.  Given "3" or
 * "3-", return {3,-1}.  Given "+4" return {0,4}.
 */
static int
parse_range (offsets, opt)
     off_t offsets[];
     char const *opt;
{
  char *o = (char *)opt;
  offsets[0] = scan_off_t(&o);
  offsets[1] = -1;

  switch(*(o++)) {

  case '\0':
    break;

  case '-': case ',':
    if (*o)
      offsets[1] = scan_off_t(&o) - offsets[0] + 1;
    break;

  case '+':
    if (*o)
      offsets[1] = scan_off_t(&o);
    break;

  default:
    return -1;		/* parse error */

  }
  return 0;		/* success */
}
